tests/helper/ssh.go

```diff
@@ -0,0 +1,290 @@
+/**
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+// Simplifies work with ssh testing
+package helper
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"unsafe"
+
+	"github.com/creack/pty"
+	sshd "github.com/gliderlabs/ssh"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// Base ssh server with no handler
+func TestSSHServer(t *testing.T, sshSrv *sshd.Server, user, pass, key string) string {
+	if pass != "" {
+		sshSrv.SetOption(sshd.PasswordAuth(func(ctx sshd.Context, password string) bool {
+			return ctx.User() == user && password == pass
+		}))
+	}
+	if key != "" {
+		sshSrv.SetOption(sshd.PublicKeyAuth(func(ctx sshd.Context, inkey sshd.PublicKey) bool {
+			return ctx.User() == user && key == string(ssh.MarshalAuthorizedKey(inkey))
+		}))
+	}
+
+	sshListener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("Unable to start SSH server to listen: %v", err)
+	}
+	t.Cleanup(func() {
+		sshListener.Close()
+	})
+	sshSrv.Addr = sshListener.Addr().String()
+
+	_, port, err := net.SplitHostPort(sshListener.Addr().String())
+	if err != nil {
+		t.Fatalf("Unable to get SSH listening port: %v", err)
+	}
+
+	go sshSrv.Serve(sshListener)
+
+	t.Log("Started Test SSH server on", sshSrv.Addr)
+
+	return port
+}
+
+func TestSSHPtyServer(t *testing.T, user, pass, key string) string {
+	sshSrv := &sshd.Server{Handler: func(s sshd.Session) {
+		t.Log("Test SSH server: handling session")
+		cmd := exec.Command("sh")
+		ptyReq, winCh, isPty := s.Pty()
+		if isPty {
+			t.Log("Test SSH server: pty is requested")
+			cmd.Env = append(cmd.Env, fmt.Sprintf("TERM=%s", ptyReq.Term))
+			f, err := pty.Start(cmd)
+			if err != nil {
+				panic(err)
+			}
+			go func() {
+				for win := range winCh {
+					setWinsize(f, win.Width, win.Height)
+				}
+			}()
+			go func() {
+				io.Copy(f, s) // stdin
+			}()
+			io.Copy(s, f) // stdout
```

Comment:
Probably obvious ... why doesn't this one (`stdout`) get wrapped in a `go func()`?

Reply:
Because it's not needed to be wrapped - it will execute together with the reverse one and block the pipe until the buffer will be closed. Goroutines are needed to execute in parallel with the main thread and here we need just one additional thread to do both things in parallel.